nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,848 @@
+        self.mims = mims if mims is not None else []
+        self.metadata = meta
+
+    @property
+    def metadata(self):
+        return self._meta
+
+    @metadata.setter
+    def metadata(self, meta):
+        """ Set the metadata for this Cifti2Header
+
+        Parameters
+        ----------
+        meta : Cifti2MetaData
+
+        Returns
+        -------
+        None
+        """
+        if meta is not None and not isinstance(meta, Cifti2MetaData):
+            raise TypeError("Not a valid Cifti2MetaData instance")
+        self._meta = meta
+
+    def add_cifti_matrix_indices_map(self, mim):
+        """ Adds a matrix indices map to the Cifti2Matrix
+
+        Parameters
+        ----------
+        mim : Cifti2MatrixIndicesMap
+        """
+        if isinstance(mim, Cifti2MatrixIndicesMap):
+            self.mims.append(mim)
+        else:
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+
+    def remove_cifti2_matrix_indices_map(self, ith):
+        """ Removes the ith matrix indices map element from the Cifti2Matrix """
+        self.mims.pop(ith)
+
+    def _to_xml_element(self):
+        assert len(self.mims) != 0 or self.metadata is not None
+        mat = xml.Element('Matrix')
+        if self.metadata:
+            mat.append(self.metadata._to_xml_element())
+        for mim in self.mims:
+            mat.append(mim._to_xml_element())
+        return mat
+
+
+class Cifti2Header(FileBasedHeader, xml.XmlSerializable):
+    ''' Class for Cifti2 header extension '''
+
+    # version = str
+
+    def __init__(self, matrix=None, version="2.0"):
+        FileBasedHeader.__init__(self)
+        xml.XmlSerializable.__init__(self)
+        if matrix is None:
+            self.matrix = Cifti2Matrix()
+        else:
+            self.matrix = matrix
+        self.version = version
+
+    def _to_xml_element(self):
+        cifti = xml.Element('CIFTI')
+        cifti.attrib['Version'] = str(self.version)
+        mat_xml = self.matrix._to_xml_element()
+        if mat_xml is not None:
+            cifti.append(mat_xml)
+        return cifti
+
+    @classmethod
+    def from_header(klass, header=None):
+        if header is None:
+            return klass()
+        if type(header) == klass:
+            return header.copy()
+        raise ValueError('header is not a Cifti2Header')
+
+    @classmethod
+    def may_contain_header(klass, binaryblock):
+        from .parse_cifti2_fast import _Cifti2AsNiftiHeader
+        return _Cifti2AsNiftiHeader.may_contain_header(binaryblock)
+
+
+class Cifti2Image(FileBasedImage):
+    # It is a Nifti2Image, but because Nifti2Image object
+    # contains both the *format* and the assumption that it's
+    # a spatial image, we can't inherit directly.
+    header_class = Cifti2Header
+    valid_exts = Nifti2Image.valid_exts
+    files_types = Nifti2Image.files_types
+    makeable = False
+    rw = True
+
+    def __init__(self, data=None, header=None, nifti_header=None):
```

Comment:
`data` in signature doesn't match `dataobj` in docstring.   `dataobj` better, for compatibility with other image types.